src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/Serializer/OtlpProtoTagWriter.cs

```diff
@@ -0,0 +1,220 @@
+    {
+        [ThreadStatic]
+        private static byte[]? threadBuffer;
+
+        public override OtlpTagWriterArrayState BeginWriteArray()
+        {
+            threadBuffer ??= new byte[2048];
+
+            return new OtlpTagWriterArrayState
+            {
+                Buffer = threadBuffer,
+                WritePosition = 0,
+            };
+        }
+
+        public override void WriteNullValue(ref OtlpTagWriterArrayState state)
+        {
+        }
+
+        public override void WriteIntegralValue(ref OtlpTagWriterArrayState state, long value)
+        {
+            var size = ComputeVarint64Size((ulong)value) + 1;
+            state.WritePosition = ProtobufSerializer.WriteTagAndLength(state.Buffer, state.WritePosition, size, FieldNumberConstants.ArrayValue_Value, ProtobufWireType.LEN);
+            state.WritePosition = ProtobufSerializer.WriteInt64WithTag(state.Buffer, state.WritePosition, FieldNumberConstants.AnyValue_Int_Value, (ulong)value);
+        }
+
+        public override void WriteFloatingPointValue(ref OtlpTagWriterArrayState state, double value)
+        {
+            state.WritePosition = ProtobufSerializer.WriteTagAndLength(state.Buffer, state.WritePosition, 9, FieldNumberConstants.ArrayValue_Value, ProtobufWireType.LEN);
+            state.WritePosition = ProtobufSerializer.WriteDoubleWithTag(state.Buffer, state.WritePosition, FieldNumberConstants.AnyValue_Double_Value, value);
+        }
+
+        public override void WriteBooleanValue(ref OtlpTagWriterArrayState state, bool value)
+        {
+            state.WritePosition = ProtobufSerializer.WriteTagAndLength(state.Buffer, state.WritePosition, 2, FieldNumberConstants.ArrayValue_Value, ProtobufWireType.LEN);
+            state.WritePosition = ProtobufSerializer.WriteBoolWithTag(state.Buffer, state.WritePosition, FieldNumberConstants.AnyValue_Bool_Value, value);
+        }
+
+        public override void WriteStringValue(ref OtlpTagWriterArrayState state, ReadOnlySpan<char> value)
+        {
+            // Write KeyValue.Value tag, length and value.
+#if NETFRAMEWORK || NETSTANDARD2_0
+            int numberOfUtf8CharsInString;
+            unsafe
+            {
+                fixed (char* p = value)
+                {
+                    numberOfUtf8CharsInString = Encoding.UTF8.GetByteCount(p, value.Length);
+                }
+            }
+#else
+            int numberOfUtf8CharsInString = Encoding.UTF8.GetByteCount(value);
+#endif
+
+            // length = numberOfUtf8CharsInString + tagSize + length field size.
+            state.WritePosition = ProtobufSerializer.WriteTagAndLength(state.Buffer, state.WritePosition, numberOfUtf8CharsInString + 2, FieldNumberConstants.ArrayValue_Value, ProtobufWireType.LEN);
+            state.WritePosition = ProtobufSerializer.WriteStringWithTag(state.Buffer, state.WritePosition, FieldNumberConstants.AnyValue_String_Value, value);
```

Comment:
This `ProtobufSerializer.WriteStringWithTag` call is going to also call `Encoding.UTF8.GetByteCount` IIRC. Could we make an overload where you can pass in the known UTF8 byte count to avoid that recalculation where we have already performed it? Or maybe make a helper which combines these `WriteTagAndLength` + `WriteStringWithTag` calls? 🤷

Reply:
I'm thinking about the same, had a similar situation when working on Span serialization. My plan is to separate the length and write part.